rego-templates/embedded.go

```diff
@@ -0,0 +1,101 @@
+package rego_templates
+
+import (
+	"bufio"
+	"embed"
+	"io/fs"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/aquasecurity/postee/v2/data"
+	"github.com/aquasecurity/postee/v2/log"
+)
+
+const (
+	COMMON_DIR               = "common"
+	LOCAL_DIR                = "."
+	REGO_PACKAGE_DECLERATION = "package"
+)
+
+var (
+	//go:embed *.rego
+	EmbeddedFiles embed.FS
+	templates     = make(map[string]string)
+	templatesMu   sync.Mutex
+
+	//go:embed common
+	EmbeddedCommonFiles embed.FS
+	commonTemplates     = make(map[string]string)
+	commonMu            sync.Mutex
+)
+
+func EmbeddedTemplates() map[string]string {
+	templatesMu.Lock()
+	defer templatesMu.Unlock()
+	if len(templates) != 0 {
+		return templates
+	}
+	populateFS(EmbeddedFiles, templates, LOCAL_DIR)
+
+	return templates
+}
+
+func EmbeddedCommon() map[string]string {
+	commonMu.Lock()
+	defer commonMu.Unlock()
+	if len(commonTemplates) != 0 {
+		return commonTemplates
+	}
+	populateFS(EmbeddedCommonFiles, commonTemplates, COMMON_DIR)
+
+	return commonTemplates
+}
+
+func populateFS(files embed.FS, storage map[string]string, dirPath string) {
+	dir, err := fs.ReadDir(files, dirPath)
+	if err != nil {
+		log.Logger.Errorf("failed to read embedded files: %s", err)
+		return
+	}
+
+	for _, file := range dir {
+		if file.IsDir() {
+			continue
+		}
+
+		bt, err := fs.ReadFile(files, filepath.Join(dirPath, file.Name()))
+		if err != nil {
+			log.Logger.Errorf("failed to read embedded file '%s': %s", file.Name(), err)
+			continue
+		}
+
+		storage[file.Name()] = string(bt)
+	}
+}
+
+func GetAllTemplates() []data.Template {
+	var templates []data.Template
+	embedded := EmbeddedTemplates()
+	for name, file := range embedded {
+		scanner := bufio.NewScanner(strings.NewReader(file))
+		for scanner.Scan() {
+			line := scanner.Text()
+			if strings.HasPrefix(line, REGO_PACKAGE_DECLERATION+" ") {
+				s := strings.Split(line, " ")
```

Comment:
technically there are cases where it won't work correctly
https://play.openpolicyagent.org/p/ivEJHEkFn2

Reply:
Agree, do you have an idea how we can approach this issue?

Reply:
I don't - it's just fun fact